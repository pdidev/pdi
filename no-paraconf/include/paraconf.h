/*******************************************************************************
* Copyright (C) 2025 Commissariat a l'energie atomique et aux energies alternatives (CEA)
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* * Redistributions of source code must retain the above copyright
*   notice, this list of conditions and the following disclaimer.
* * Redistributions in binary form must reproduce the above copyright
*   notice, this list of conditions and the following disclaimer in the
*   documentation and/or other materials provided with the distribution.
* * Neither the name of CEA nor the names of its contributors may be used to
*   endorse or promote products derived from this software without specific
*   prior written permission.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
******************************************************************************/

/** \file paraconf.h for deactivation
 */

#ifndef PARACONF_H_
#define PARACONF_H__H_

#ifdef __cplusplus
extern "C" {
#endif

/** \addtogroup error
 * \{
 */

typedef struct PC_tree_s {
} PC_tree_t;
struct _IO_FILE;
typedef struct _IO_FILE FILE;

static inline PC_tree_t PC_parse_string(const char* document)
{
	PC_tree_t mock = {};
	return mock;
}

static inline PC_tree_t PC_parse_file(FILE *file)
{
	PC_tree_t mock = {};
	return mock;
}

#ifdef __cplusplus
} // extern C
#endif

#endif // PARACONF_H_
