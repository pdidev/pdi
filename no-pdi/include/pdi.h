/*******************************************************************************
* Copyright (C) 2025 Commissariat a l'energie atomique et aux energies alternatives (CEA)
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* * Redistributions of source code must retain the above copyright
*   notice, this list of conditions and the following disclaimer.
* * Redistributions in binary form must reproduce the above copyright
*   notice, this list of conditions and the following disclaimer in the
*   documentation and/or other materials provided with the distribution.
* * Neither the name of CEA nor the names of its contributors may be used to
*   endorse or promote products derived from this software without specific
*   prior written permission.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
******************************************************************************/

/** \file pdi.h for deactivation
 *
 * C user API
 *
 * The user facing API is the interface offered by PDI to C application
 * developers.
 *
 * \defgroup init_final Initialization and finalization
 *
 * The initialization and finalization part of the API is used to setup PDI,
 * release its resources and check version information.
 *
 * \defgroup annotation Code annotation
 *
 * The code annotation API is the main interface to use in the code.
 *
 * It offers functions that can be called from code with no side effect by
 * default and that can therefore be considered as annotations.
 *
 * \defgroup error Error handling
 *
 * The error handling API supports checking the error status of PDI.
 *
 * By default, errors in PDI C API are signaled by a return code of type
 * PDI_status_t and an error message can be retrieved with the PDI_errmsg
 * function. This default behavior can be changed by replacing the error handler
 * with the PDI_errhandler function.
 *
 */

#ifndef PDI_H_
#define PDI_H_

#ifdef __cplusplus
extern "C" {
#endif

/** \addtogroup error
 * \{
 */

/** Error codes of PDI
 */
typedef enum PDI_status_e {
	/// everything went well
	PDI_OK = 0,
	/// on an input call, no such data is available
	PDI_UNAVAILABLE,
	/// The configuration file is invalid
	PDI_ERR_CONFIG,
	/// A value expression is invalid
	PDI_ERR_VALUE,
	/// Tried to load a non-existing plugin
	PDI_ERR_PLUGIN,
	/// Implementation limitation (typically an unimplemented feature)
	PDI_ERR_IMPL,
	/// A system error occured (OS, etc.)
	PDI_ERR_SYSTEM,
	/** A call to a function has been made at a wrong time (e.g. closing an
	 *  unopened transaction)
	 */
	PDI_ERR_STATE,
	/// A conflict of onwership over a content has been raised
	PDI_ERR_RIGHT,
	/// Invalid type error
	PDI_ERR_TYPE

} PDI_status_t;

/** Type of a callback function used when an error occurs
 * \param status the error code
 * \param message the human-readable error message
 * \param context a user-provided context
 */
typedef void (*PDI_errfunc_f)(PDI_status_t status, const char* message, void* context);

/** Definition of an error handler
 */
typedef struct PDI_errhandler_s {
	/// The function to handle the error (none if NULL)
	PDI_errfunc_f func;

	/// the context that will be provided to the function
	void* context;

} PDI_errhandler_t;

#ifndef PDI_NULL_HANDLER_DEFINED
#define PDI_NULL_HANDLER_DEFINED
/** Does nothing
 */
const PDI_errhandler_t PDI_NULL_HANDLER = {NULL, NULL};

/** Prints the error message and aborts if the status is invalid
 */
const PDI_errhandler_t PDI_ASSERT_HANDLER = PDI_NULL_HANDLER;

/** Prints the error message and continue if the status is invalid
 */
const PDI_errhandler_t PDI_WARN_HANDLER = PDI_NULL_HANDLER;
#endif

/** Return a human-readabe message describing the last error that occured in PDI
 */
static inline const char* PDI_errmsg(void)
{
	return "";
}

/** Sets the error handler to use
 */
static inline PDI_errhandler_t PDI_errhandler(PDI_errhandler_t)
{
	return PDI_NULL_HANDLER;
}

/// \}

/** \addtogroup init_final Initialization and finalization
 *
 * The initialization and finalization part of the API is used to setup PDI,
 * release its resources and check version information.
 * \{
 */

/** Initializes PDI
 */
static inline PDI_status_t PDI_init(PC_tree_t)
{
	return PDI_OK;
}

/** Finalizes PDI
 * \return an error status
 */
static inline PDI_status_t PDI_finalize(void)
{
	return PDI_OK;
}

/** Checks PDI API version
 */
static inline PDI_status_t PDI_version(unsigned long*, unsigned long)
{
	return PDI_OK;
}

/// \}

/** \addtogroup annotation
 * \{
 */

/**
 * Access directions
 */
typedef enum PDI_inout_e {
	/// No data transfert
	PDI_NONE = 0,
	/// data tranfer from PDI to the main code
	PDI_IN = 1,
	/// data transfer from the main code to PDI
	PDI_OUT = 2,
	/// data transfer in both direction
	PDI_INOUT = 3

} PDI_inout_t;

/** Shares some data with PDI. The user code should not modify it before
 * a call to either PDI_release or PDI_reclaim.
 */
static inline PDI_status_t PDI_share(const char*, void*, PDI_inout_t)
{
	return PDI_OK;
}

/** Requests for PDI to access a data buffer.
 */
static inline PDI_status_t PDI_access(const char*, void**, PDI_inout_t)
{
	return PDI_OK;
}

/** Releases ownership of a data shared with PDI. PDI is then responsible to
 * free the associated memory whenever necessary.
 */
static inline PDI_status_t PDI_release(const char*)
{
	return PDI_OK;
}

/** Reclaims ownership of a data buffer shared with PDI. PDI does not manage
 * the buffer memory anymore.
 */
static inline PDI_status_t PDI_reclaim(const char*)
{
	return PDI_OK;
}

/** Triggers a PDI "event"
 */
static inline PDI_status_t PDI_event(const char*)
{
	return PDI_OK;
}

/** Shortly exposes some data to PDI. Equivalent to PDI_share + PDI_reclaim.
 */
static inline PDI_status_t PDI_expose(const char*, void*, PDI_inout_t)
{
	return PDI_OK;
}

/** Performs multiple exposes at once. All the data is shared in order they were specified
 *  and reclaimed in reversed order after an event is triggered.
 *
 *  NULL argument indicates an end of the list.
 */
static inline PDI_status_t PDI_multi_expose(const char*, const char*, void*, PDI_inout_t, ...)
{
	return PDI_OK;
}

#ifdef PDI_WITH_DEPRECATED

/** Begin a transaction in which all PDI_expose calls are grouped.
 *
 * This requires a call to PDI_transaction_end to close the transaction.
 *
 * \deprecated the transaction part of the API is deprecated, the
 * PDI_multi_expose function should be used instead.
 *
 * \see PDI_expose the function used to expose data inside the transaction
 * \see PDI_transaction_end the function used to end the transaction
 *
 * \return an error status
 */
static inline PDI_status_t PDI_transaction_begin(const char*)
{
	return PDI_OK;
}

/** Ends the previously opened transaction.
 *
 * \deprecated the transaction part of the API is deprecated, the
 * PDI_multi_expose function should be used instead.
 *
 * \see PDI_transaction_begin the function used to start the transaction
 * \see PDI_expose the function used to expose data inside the transaction
 *
 * \return an error status
 */
static inline PDI_status_t PDI_transaction_end(void)
{
	return PDI_OK;
}

#endif // PDI_WITH_DEPRECATED

/// \}

#ifdef __cplusplus
} // extern C
#endif


#endif // PDI_H_
