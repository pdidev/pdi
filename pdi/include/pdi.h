/*******************************************************************************
* Copyright (C) 2015-2021 Commissariat a l'energie atomique et aux energies alternatives (CEA)
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* * Redistributions of source code must retain the above copyright
*   notice, this list of conditions and the following disclaimer.
* * Redistributions in binary form must reproduce the above copyright
*   notice, this list of conditions and the following disclaimer in the
*   documentation and/or other materials provided with the distribution.
* * Neither the name of CEA nor the names of its contributors may be used to
*   endorse or promote products derived from this software without specific
*   prior written permission.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
******************************************************************************/

/** \file pdi.h
 *
 * C user API
 *
 * The user facing API is the interface offered by PDI to C application
 * developers.
 *
 * \defgroup init_final Initialization and finalization
 *
 * The initialization and finalization part of the API is used to setup PDI,
 * release its resources and check version information.
 *
 * \defgroup annotation Code annotation
 *
 * The code annotation API is the main interface to use in the code.
 *
 * It offers functions that can be called from code with no side effect by
 * default and that can therefore be considered as annotations.
 *
 * \defgroup error Error handling
 *
 * The error handling API supports checking the error status of PDI.
 *
 * By default, errors in PDI C API are signaled by a return code of type
 * PDI_status_t and an error message can be retrieved with the PDI_errmsg
 * function. This default behavior can be changed by replacing the error handler
 * with the PDI_errhandler function.
 * 
 */

#ifndef PDI_H_
#define PDI_H_

#include <paraconf.h>

#include <pdi/export.h>
#include <pdi/version.h>

//#include "/gpfs/softs/spack_0.17/opt/spack/linux-centos7-cascadelake/gcc-11.2.0/openmpi-4.1.1-ujlwnrlh5sewm2rxkpio3h5mariwgetn/include/mpi.h"

#ifdef __cplusplus
extern "C" {
#endif

/** \addtogroup error
 * \{
 */

/** Error codes of PDI
 */
typedef enum PDI_status_e {
	/// everything went well
	PDI_OK = 0,
	/// on an input call, no such data is available
	PDI_UNAVAILABLE,
	/// The configuration file is invalid
	PDI_ERR_CONFIG,
	/// A value expression is invalid
	PDI_ERR_VALUE,
	/// Tried to load a non-existing plugin
	PDI_ERR_PLUGIN,
	/// Implementation limitation (typically an unimplemented feature)
	PDI_ERR_IMPL,
	/// A system error occured (OS, etc.)
	PDI_ERR_SYSTEM,
	/** A call to a function has been made at a wrong time (e.g. closing an
	 *  unopened transaction)
	 */
	PDI_ERR_STATE,
	/// A conflict of onwership over a content has been raised
	PDI_ERR_RIGHT,
	/// Invalid type error
	PDI_ERR_TYPE

} PDI_status_t;

/** Type of a callback function used when an error occurs
 * \param status the error code
 * \param message the human-readable error message
 * \param context a user-provided context
 */
typedef void (*PDI_errfunc_f)(PDI_status_t status, const char* message, void* context);

/** Definition of an error handler
 */
typedef struct PDI_errhandler_s {
	/// The function to handle the error (none if NULL)
	PDI_errfunc_f func;

	/// the context that will be provided to the function
	void* context;

} PDI_errhandler_t;

/** Prints the error message and aborts if the status is invalid
 */
extern const PDI_errhandler_t PDI_EXPORT PDI_ASSERT_HANDLER;

/** Prints the error message and continue if the status is invalid
 */
extern const PDI_errhandler_t PDI_EXPORT PDI_WARN_HANDLER;

/** Does nothing
 */
extern const PDI_errhandler_t PDI_EXPORT PDI_NULL_HANDLER;


/** Return a human-readabe message describing the last error that occured in PDI
 */
const char PDI_EXPORT* PDI_errmsg(void);

/** Sets the error handler to use
 *
 * PDI_asserthandler is the default handler before this function is called
 *
 * \param handler the new handler to set
 * \return the previous handler
 */
PDI_errhandler_t PDI_EXPORT PDI_errhandler(PDI_errhandler_t handler);

/// \}

/** \addtogroup init_final Initialization and finalization
 *
 * The initialization and finalization part of the API is used to setup PDI,
 * release its resources and check version information.
 * \{
 */

/** Initializes PDI
 * \param[in] conf the configuration
 * \return an error status
 */
//PDI_status_t PDI_EXPORT PDI_init(PC_tree_t conf,const char* configfile, int *is_client,MPI_Comm *damaris_communicator);
PDI_status_t PDI_EXPORT PDI_init(PC_tree_t conf,const char* configfile, int *is_client);


/** Finalizes PDI
 * \return an error status
 */
PDI_status_t PDI_EXPORT PDI_finalize(int);

//PDI_status_t PDI_EXPORT PDI_finalize_Damaris(int *is_client);

/** Checks PDI API version
 *
 * \param[out] provided version if non-null it is filled with the provided API version
 * \param[in] expected if non-zero the expected API version
 * \return an error status if the expected version is incompatible with the
 * provided one
 */
PDI_status_t PDI_EXPORT PDI_version(unsigned long* provided, unsigned long expected);

/// \}

/** \addtogroup annotation
 * \{
 */

/**
 * Access directions
 */
typedef enum PDI_inout_e {
	/// No data transfert
	PDI_NONE = 0,
	/// data tranfer from PDI to the main code
	PDI_IN = 1,
	/// data transfer from the main code to PDI
	PDI_OUT = 2,
	/// data transfer in both direction
	PDI_INOUT = 3

} PDI_inout_t;

/** Shares some data with PDI. The user code should not modify it before
 * a call to either PDI_release or PDI_reclaim.
 * \param[in] name the data name
 * \param[in,out] data the accessed data
 * \param[in] access whether the data can be accessed for read or write
 *                   by PDI
 * \return an error status
 * \pre the user code owns the data buffer
 * \post ownership of the data buffer is shared between PDI and the user code
 *
 * the access parameter is a binary OR of PDI_IN & PDI_OUT.
 * * PDI_IN means PDI can set the buffer content
 * * PDI_OUT means the buffer contains data that can be accessed by PDI
 */
PDI_status_t PDI_EXPORT PDI_share(const char* name, void* data, PDI_inout_t access);

/** Requests for PDI to access a data buffer.
 * \param[in] name the data name
 * \param[in,out] buffer a pointer to the accessed data buffer
 * \param[in] inout the access properties (PDI_IN, PDI_OUT, PDI_INOUT)
 * \return an error status
 * \pre PDI owns the data buffer
 * \post ownership of the data buffer is shared between PDI and the user code
 */
PDI_status_t PDI_EXPORT PDI_access(const char* name, void** buffer, PDI_inout_t inout);

/** Releases ownership of a data shared with PDI. PDI is then responsible to
 * free the associated memory whenever necessary.
 * \param[in] name name of the data to release
 * \return an error status
 * \pre ownership of the data buffer is shared between PDI and the user code
 * \pre PDI owns the data buffer
 */
PDI_status_t PDI_EXPORT PDI_release(const char* name);

/** Reclaims ownership of a data buffer shared with PDI. PDI does not manage
 * the buffer memory anymore.
 * \param[in] name name of the data to reclaim
 * \return an error status
 * \pre ownership of the data buffer is shared between PDI and the user code
 * \post the user code owns the data buffer
 */
PDI_status_t PDI_EXPORT PDI_reclaim(const char* name);

/** Triggers a PDI "event"
 * \param[in] event the event name
 * \return an error status
 */
PDI_status_t PDI_EXPORT PDI_event(const char* event);

/** Shortly exposes some data to PDI. Equivalent to PDI_share + PDI_reclaim.
 * \param[in] name the data name
 * \param[in] data the exposed data
 * \param[in] access whether the data can be accessed for read or write
 *                   by PDI
 * \return an error status
 */
PDI_status_t PDI_EXPORT PDI_expose(const char* name, void* data, PDI_inout_t access);

/** Performs multiple exposes at once. All the data is shared in order they were specified
 *  and reclaimed in reversed order after an event is triggered.
 *
 *  NULL argument indicates an end of the list.
 *
 * \param[in] event_name the name of the event that will be triggered when
 *                       all data become available
 * \param[in] name the data name
 * \param[in] data the exposed data
 * \param[in] access whether the data can be accessed for read or write by PDI
 * \param[in] ... (additional arguments) additional list of data to expose,
 *                each should contain name, data and access, NULL argument
 *                inidactes an end of the list.
 * \return an error status
 */
PDI_status_t PDI_EXPORT PDI_multi_expose(const char* event_name, const char* name, void* data, PDI_inout_t access, ...);

#ifdef PDI_WITH_DEPRECATED

/** Begin a transaction in which all PDI_expose calls are grouped.
 *
 * This requires a call to PDI_transaction_end to close the transaction.
 *
 * \deprecated the transaction part of the API is deprecated, the
 * PDI_multi_expose function should be used instead.
 *
 * \see PDI_expose the function used to expose data inside the transaction
 * \see PDI_transaction_end the function used to end the transaction
 *
 * \param[in] name the name of the transaction (an event thus named will be
 *                 triggered when all data become available)
 * \return an error status
 */
PDI_status_t PDI_DEPRECATED_EXPORT PDI_transaction_begin(const char* name);

/** Ends the previously opened transaction.
 *
 * \deprecated the transaction part of the API is deprecated, the
 * PDI_multi_expose function should be used instead.
 *
 * \see PDI_transaction_begin the function used to start the transaction
 * \see PDI_expose the function used to expose data inside the transaction
 *
 * \return an error status
 */
PDI_status_t PDI_DEPRECATED_EXPORT PDI_transaction_end(void);

#endif // PDI_WITH_DEPRECATED

/// \}

#ifdef __cplusplus
} // extern C
#endif


#endif // PDI_H_
