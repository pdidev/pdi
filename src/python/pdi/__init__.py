#*******************************************************************************
# Copyright (C) 2015-2019 Commissariat a l'energie atomique et aux energies alternatives (CEA)
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
# * Redistributions of source code must retain the above copyright
#   notice, this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright
#   notice, this list of conditions and the following disclaimer in the
#   documentation and/or other materials provided with the distribution.
# * Neither the name of CEA nor the names of its contributors may be used to
#   endorse or promote products derived from this software without specific
#   prior written permission.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#*****************************************************************************/

from ._pdi import access, Error, event, finalize, init, reclaim, release, version, OUT, IN, INOUT, NONE
import pdi._pdi
import numpy as np
import inspect

def share(name, data, access):
    if (isinstance(data, np.ndarray)):
        data_np_array = data
    elif (access == OUT or access == NONE):
        # data is not numpy array
        try:
            data_np_array = np.array(data)
        except:
            raise Error("`" + name + "' share: Type is not supported by PDI, cannot insert it into numpy array")
    else:
        raise Error("`" + name + "' share: IN and INOUT can be only done with numpy array data type")
    pdi._pdi.share(name, data_np_array, access)

def expose(name, data, access):
    share(name, data, access)
    reclaim(name)

def multi_expose(event_name, expose_list):
    exposed = []
    try:
        for (name, data, access) in expose_list:
            share(name, data, access)
            exposed.append(name)
        event(event_name)
    except:
        pass
    final_error = ()
    for name in exposed:
        try:
            reclaim(name)
        except Exception as e:
            final_error += (e)
    if (final_error != ()):
        raise final_error
