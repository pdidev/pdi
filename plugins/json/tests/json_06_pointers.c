/*******************************************************************************
 * Copyright (C) 2023-2024 Commissariat a l'energie atomique et aux energies alternatives (CEA)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * * Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the distribution.
 * * Neither the name of CEA nor the names of its contributors may be used to
 *   endorse or promote products derived from this software without specific
 *   prior written permission.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

#include <paraconf.h>
#include <stdint.h>
#include <stdio.h>
#include <unistd.h>
#include <pdi.h>

int main()
{
	const char* CONFIG_YAML
		= "pdi:                                                        \n"
		  "  data:                                                     \n"
		  "    pvar_uint16_t: {type: pointer, subtype: uint16_t}       \n"
		  "    pvar_uint32_t: {type: pointer, subtype: uint32_t}       \n"
		  "    pvar_uint64_t: {type: pointer, subtype: uint64_t}       \n"
		  "    pvar_int8_t: {type: pointer, subtype: int8_t}           \n"
		  "    pvar_int16_t: {type: pointer, subtype: int16_t}         \n"
		  "    pvar_int32_t: {type: pointer, subtype: int32_t}         \n"
		  "    pvar_int64_t: {type: pointer, subtype: int64_t}         \n"
		  "    pvar_float: {type: pointer, subtype: float}             \n"
		  "    pvar_double: {type: pointer, subtype: double}           \n"
		  "    pvar_char: {type: pointer, subtype: char}               \n"
		  "                                                            \n"
		  "    pvar_string:                                            \n"
		  "      type: pointer                                         \n"
		  "      subtype: { type: array, subtype: char, size: 8 }      \n"
		  "    pvar_array:                                             \n"
		  "      type: pointer                                         \n"
		  "      subtype : { type: array, subtype: int32_t, size: 3 }  \n"
		  "    pvar_record:                                            \n"
		  "      type: pointer                                         \n"
		  "      subtype:                                              \n"
		  "        type: struct                                        \n"
		  "        members:                                            \n"
		  "            - var_int32_t1: int32_t                         \n"
		  "            - var_int32_t2: int32_t                         \n"
		  "            - var_char: char                                \n"
		  "            - var_double: double                            \n"
		  "                                                            \n"
		  "    var_int_pointer_array:                                  \n"
		  "      type: array                                           \n"
		  "      size: 2                                               \n"
		  "      subtype:                                              \n"
		  "        type: pointer                                       \n"
		  "        subtype: {type: array, size: 3, subtype: int32_t}   \n"
		  "                                                            \n"
		  "  plugins:                                                  \n"
		  "    json:                                                   \n"
		  "      - file: json_06_pointers.json                         \n"
		  "        write: [pvar_uint16_t, pvar_uint32_t, pvar_uint64_t, pvar_int8_t, pvar_int16_t, pvar_int32_t, pvar_int64_t, pvar_float, "
		  "pvar_double, pvar_char, pvar_array, pvar_string, pvar_record, var_int_pointer_array]\n";

	PC_tree_t conf = PC_parse_string(CONFIG_YAML);
	PDI_init(PC_get(conf, ".pdi"));

	// Scalars
	uint16_t _var_uint16_t = 2;
	uint32_t _var_uint32_t = 3;
	uint64_t _var_uint64_t = 4;

	int8_t _var_int8_t = 5;
	int16_t _var_int16_t = 6;
	int32_t _var_int32_t = 7;
	int64_t _var_int64_t = 8;
	float _var_float = 9;
	double _var_double = 10;

	char _var_char = 'a';

	uint16_t* _p_var_uint16_t = &_var_uint16_t;
	uint32_t* _p_var_uint32_t = &_var_uint32_t;
	uint64_t* _p_var_uint64_t = &_var_uint64_t;

	int8_t* _p_var_int8_t = &_var_int8_t;
	int16_t* _p_var_int16_t = &_var_int16_t;
	int32_t* _p_var_int32_t = &_var_int32_t;
	int64_t* _p_var_int64_t = &_var_int64_t;
	float* _p_var_float = &_var_float;
	double* _p_var_double = &_var_double;

	char* _p_var_char = &_var_char;

	PDI_expose("pvar_uint16_t", &_p_var_uint16_t, PDI_OUT);
	PDI_expose("pvar_uint32_t", &_p_var_uint32_t, PDI_OUT);
	PDI_expose("pvar_uint64_t", &_p_var_uint64_t, PDI_OUT);

	PDI_expose("pvar_int8_t", &_p_var_int8_t, PDI_OUT);
	PDI_expose("pvar_int16_t", &_p_var_int16_t, PDI_OUT);
	PDI_expose("pvar_int32_t", &_p_var_int32_t, PDI_OUT);
	PDI_expose("pvar_int64_t", &_p_var_int64_t, PDI_OUT);
	PDI_expose("pvar_float", &_p_var_float, PDI_OUT);
	PDI_expose("pvar_double", &_p_var_double, PDI_OUT);

	PDI_expose("pvar_char", &_p_var_char, PDI_OUT);

	/* STRING */
	char _var_string[] = "string12";
	char* _p_var_string = _var_string;
	PDI_expose("pvar_string", &_p_var_string, PDI_OUT);

	/* ARRAY */
	int32_t _var_array[3] = {11, 12, 13};
	int32_t* _p_var_array = _var_array;
	PDI_expose("pvar_array", &_p_var_array, PDI_OUT);

	/* RECORD */
	struct record_t {
		int32_t _var_int32t1;
		int32_t _var_int32t2;
		char _var_a; // padding
		double _var_double;
	};
	struct record_t _var_record1 = {14, 15, 'a', 16.};
	struct record_t* _p_var_record1 = &_var_record1;

	PDI_expose("pvar_record", &_p_var_record1, PDI_OUT);

	/* ARRAYS */
	int32_t _var_array1[3] = {1, 2, 3};
	int32_t _var_array2[3] = {4, 5, 6};
	int32_t* _var_int_pointer_array[2] = {_var_array1, _var_array2};
	PDI_expose("var_int_pointer_array", _var_int_pointer_array, PDI_OUT);

	PDI_finalize();
	return 0;
}
