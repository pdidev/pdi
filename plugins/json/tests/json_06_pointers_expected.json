[
{"pvar_uint16_t": 2},
{"pvar_uint32_t": 3},
{"pvar_uint64_t": 4},
{"pvar_int8_t": 5},
{"pvar_int16_t": 6},
{"pvar_int32_t": 7},
{"pvar_int64_t": 8},
{"pvar_float": 9.000000},
{"pvar_double": 10.000000},
{"pvar_char": "a"},
{"pvar_string": "string12"},
{"pvar_array": [
    11,
    12,
    13
  ]},
{"pvar_record": {
    "var_int32_t1": 14,
    "var_int32_t2": 15,
    "var_char": "a",
    "var_double": 16.000000
  }},
{"var_int_pointer_array": [
  [
      1,
      2,
      3
    ],
  [
      4,
      5,
      6
    ]
]}
]